File: project/skyway_android_sdk/libskywaytest/src/main/cpp/wrapper/http_client.cpp
//
//  http_client.cpp
//  skyway_android
//
//  Copyright © 2022 NTT Communications. All rights reserved.
//
#include "http_client.hpp"

#include <jni.h>

#include <boost/optional.hpp>
#include <future>
#include <json.hpp>
#include <string>

namespace skyway {
namespace network {

using Response = skyway::network::interface::HttpClient::Response;

void HttpClient::Setup(jobject j_http) { HttpClient::j_http_ = j_http; }

HttpClient::HttpClient() {
    client_ = std::make_unique<skyway_android::network::HttpClient>(HttpClient::j_http_);
}

std::future<boost::optional<Response>> HttpClient::Request(const std::string &url,
                                                           const std::string &method,
                                                           const nlohmann::json &header,
                                                           const nlohmann::json &body) {
    return client_->Request(url, method, header, body);
}

jobject HttpClient::j_http_ = nullptr;

}  // namespace network
}  // namespace skyway

File: project/skyway_android_sdk/core/src/main/java/com/ntt/skyway/core/network/HttpClient.kt
package com.ntt.skyway.core.network

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.ntt.skyway.core.util.Logger
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.io.IOException

/**
 *  @suppress
 */
object HttpClient {
    private val JSON = "application/json; charset=utf-8".toMediaType()
    private val logging = HttpLoggingInterceptor { message -> Logger.logD(message) }
    private var client: OkHttpClient

    init {
        logging.level = HttpLoggingInterceptor.Level.BASIC
        client = OkHttpClient.Builder()
            .addInterceptor(logging)
            .build()
    }

    private fun request(
        url: String,
        method: String,
        header: String,
        body: String,
        id: String
    ) {
        val requestBuilder = Request.Builder().url(url)
        val headerJson = Gson().fromJson(header, JsonObject::class.java)

        val headersBuilder = Headers.Builder()
        headerJson.keySet().forEach {
            headersBuilder.add(it, headerJson.get(it).asString)

        }
        requestBuilder.headers(headersBuilder.build())


        val request = when (method) {
            "GET" -> requestBuilder.get().build()
            "POST" -> requestBuilder.post(body.toRequestBody(JSON)).build()
            "PUT" -> requestBuilder.put(body.toRequestBody(JSON)).build()
            "DELETE" -> requestBuilder.delete().build()
            else -> throw IllegalArgumentException("Invalid method")
        }

        client.newCall(request).enqueue(object : Callback {
            override fun onResponse(call: Call, response: Response) {
                nativeOnResponse(
                    id,
                    response.code,
                    response.body?.string() ?: "{}",
                    Gson().toJson(response.headers).toString()
                )
            }

            override fun onFailure(call: Call, e: IOException) {
                nativeOnFailure(id)
            }
        })
    }

    private external fun nativeOnResponse(
        id: String,
        status: Int,
        body: String,
        headers: String
    )

    private external fun nativeOnFailure(id: String)
}
